### PR TITLE
Add performance best practices guide

### PR DESCRIPTION
**Pull request by ananda1066**
Ported performance best practices guide from prod to open-source per @jtattermusch's request.

### CODE REVIEW

File: content/en/docs/guides/performance.md
Comment (ananda1066) [inline content/en/docs/guides/performance.md:50]: @vjpai Would this apply to the OSS callback API as well?

File: content/en/docs/guides/performance.md
Comment (ananda1066) [inline content/en/docs/guides/performance.md:71]: @jtattermusch I wasn't able to find a link for the sync-with-deferred-read interface in github, any idea what I could link here?

**Comment by srini100** [comment]
Thanks for this PR! Added a few reviewers for different languages.

**Review by gnossen — APPROVED** [review]

File: content/en/docs/guides/performance.md
Comment (markdroth) [reply content/en/docs/guides/performance.md:71]: I don't think there is currently anything in OSS about this.  But I'd really love to see our generated code guide moved to OSS, since it's incredibly useful (for a lot more than just this one topic).

File: content/en/docs/guides/performance.md
Comment (markdroth) [inline content/en/docs/guides/performance.md:40]: The formatting isn't quite right here.  This note applies to this entire bullet, not just to the second work-around option.

Stepping back for a moment, though, I'm not sure it actually makes sense to cover this topic in a performance doc, since it's really more of a missing feature.  Most of the time when people run into this, it's not about performance, it's about things that just aren't working.  And I'd really rather not tell people to create multiple channels, even though we know that they sometimes need to do so today.

If we do have to mention this, let's at least point to https://github.com/grpc/grpc/issues/21386.

**Review by markdroth — CHANGES_REQUESTED** [review]

File: content/en/docs/guides/performance.md
Comment (vjpai) [reply content/en/docs/guides/performance.md:50]: Yes. It's slower than sync at OSS for the time being but we  (well, you) will improve performance and it's only slower for truly high-QPS workloads at the current moment.

Comment (vjpai) [reply content/en/docs/guides/performance.md:50]: ... and it's actually a usable API unlike CQ-based async

File: content/en/docs/guides/performance.md
Comment (vjpai) [reply content/en/docs/guides/performance.md:71]: Maybe just omit the line for now and port the guide over from Google3 in a follow-on PR? I think we've never had an actual user for this interface that we know of. It was a feature request but the project that wanted it ended up going nowhere.

File: content/en/docs/guides/performance.md
Comment (vjpai) [reply content/en/docs/guides/performance.md:40]: @markdroth how about if this moves to an advanced topic or special topic section?

Comment (vjpai) [reply content/en/docs/guides/performance.md:40]: ... and of course with a caveat that the issue is a temporary problem, this is a temporary workaround, and that it won't be needed indefinitely in the future.

File: content/en/docs/guides/performance.md
Comment (vjpai) [inline content/en/docs/guides/performance.md:62]: I feel like this is a special topic also.

File: content/en/docs/guides/performance.md
Comment (vjpai) [inline content/en/docs/guides/performance.md:92]: Maybe add a line reminding the reader that  this is unlke other languages.
```suggestion
    gRPC Python, unlike the other languages supported by gRPC.
```

**Review by vjpai — APPROVED** [review]
I already reviewed internally and have only minor comments at this point.

File: content/en/docs/guides/performance.md
Comment (ejona86) [inline content/en/docs/guides/performance.md:14]: I admit to people that steams are faster than unary RPCs, but I _discourage_ using them for performance reasons. You have be seriously hurting for performance before the costs make sense. If the unary RPCs would be _easier_ as a stream, go for it. But using a stream _just_ for the gRPC-level performance seems a mistake.

I _do_ tell people to use streams if it makes _their application_ faster, e.g., because of reduced synchronization between backends.

File: content/en/docs/guides/performance.md
Comment (ejona86) [inline content/en/docs/guides/performance.md:36]: s/randomly//. Round robin is common and isn't random. I know that isn't what this is actually meaning, but using random here may be misleading and doesn't seem to add value.

File: content/en/docs/guides/performance.md
Comment (ejona86) [inline content/en/docs/guides/performance.md:86]: The main thing here is that the executor would have a limit on number of threads to avoid excessive context switches and memory overhead under high load. So I'd say, "Provide a custom executor that limits the number of threads, based on your workload."

File: content/en/docs/guides/performance.md
Comment (ananda1066) [reply content/en/docs/guides/performance.md:14]: Hm interesting, this bullet does mention that "Streams are particularly well-suited to handling    a long-lived logical flow of data from the client-to-server,    server-to-client, or in both directions". We could emphasize this point at the start of this bullet as the main reason to use streaming RPCs?

File: content/en/docs/guides/performance.md
Comment (ananda1066) [reply content/en/docs/guides/performance.md:40]: I understand this not being so much a performance issue.. I'm fine with omitting it if you have a strong opinion @markdroth. Alternatively, I can make it an advanced topic with a link to grpc/grpc#21386. Let me know which you would prefer.

File: content/en/docs/guides/performance.md
Comment (ananda1066) [reply content/en/docs/guides/performance.md:50]: @vjpai High-QPS meaning what (a rough estimate is fine)? It seems for the OSS version, we don't want to publicize that the callback API should be favored for all RPCs at this point

File: content/en/docs/guides/performance.md
Comment (ananda1066) [reply content/en/docs/guides/performance.md:71]: I don't have bandwidth to move the code guide over at the moment, but I'll create bugs for this internally for when I can get to this. In the meantime, I'll omit this topic as it seems fairly low-priority.

**Review by ananda1066 — COMMENTED** [review]
Addressed the comments - PTAL and follow up where needed.

**Review by jtattermusch — APPROVED** [review]
Thanks for adding this! I'm happy with the contents as long as language owners are happy with the corresponding sections.

File: content/en/docs/guides/performance.md
Comment (ejona86) [reply content/en/docs/guides/performance.md:14]: Yeah, that would be fine. Mainly I'd want to downplay per-RPC overhead.

We can mention that streams have less overall overhead, but it'd be important to downplay it and mention that abusing streams might increase performance at small scale but reduce scalability due to LB issues and complexity (although obviously some rare cases are appropriate to pay those costs). Basically, I'd focus on optimizing the application architecture.

**Comment by dfawley** [comment]
What is this PR waiting for? Approval from @ejona86 and @markdroth ?

File: content/en/docs/guides/performance.md
Comment (markdroth) [inline content/en/docs/guides/performance.md:39]: Please remove the trailing backslash here.

File: content/en/docs/guides/performance.md
Comment (markdroth) [reply content/en/docs/guides/performance.md]: I'm still not hearing an argument about how this has anything to do with performance, so I'm not really convinced it belongs here.  Although to be fair, my main concern is not actually about whether or not it belongs in this particular doc; it's that I'm not really comfortable recommending that people create multiple channels, regardless of which doc we do it in.  But I guess this is the only work-around for this problem that exists today, so documenting it is fair, regardless of how unhappy it makes me. :/

So I guess this is fine.  But please do add a link to grpc/grpc#21386 for more info.

Comment (ananda1066) [reply content/en/docs/guides/performance.md]: Added link to grpc/grpc#21386.

File: content/en/docs/guides/performance.md
Comment (ananda1066) [reply content/en/docs/guides/performance.md]: Done, emphasized that streams should only be used for long-lived data flows and their impact on scalability

File: content/en/docs/guides/performance.md
Comment (markdroth) [inline content/en/docs/guides/performance.md:2]: This formats a little weirdly.  I think it might be better to just say:

```
# Performance Best Practices
```

**Review by markdroth — APPROVED** [review]

File: content/en/docs/guides/performance.md
Comment (ananda1066) [reply content/en/docs/guides/performance.md:2]: What part of the formatting is weird? I followed the other guides formatting and using `title` is how the guide appears in the navigation menu. I don't really see anything strange when I serve it locally?

Comment (markdroth) [reply content/en/docs/guides/performance.md:2]: When I did "View file" it showed up as a sort of table.  If that's expected for the structure on grpc.io, then it's fine.

Comment (ananda1066) [reply content/en/docs/guides/performance.md:2]: Oh yeah that's not how it looks in the real display. I think there's some formatting logic omitted from the rendering here.

File: content/en/docs/guides/performance.md
Comment (ejona86) [inline content/en/docs/guides/performance.md:23]: Streams do not need to be long-lived to provide benefit and unary can be long-lived. Like I mentioned before, I'd say:

> so they should only be used when they provide substantial performance or simplicity benefit to application logic. Use streams to optimize the application, not gRPC.

If we want to expand on that to make it more clear, we could give examples like "grouping related requests over time to reduce application processing per message" and "making sure a multi-part operation is handled by the same backend". But really, just "if your application benefits, consider it" seems best; I think examples would be better if we were saying cases that _didn't_ work well with streams.

**Review by ejona86 — APPROVED** [review]

File: content/en/docs/guides/performance.md
Comment (ananda1066) [reply content/en/docs/guides/performance.md]: Done, thanks!

**Pull request merged**
